spec/compiler/normalize/select_spec.cr

```diff
@@ -0,0 +1,68 @@
+require "../../spec_helper"
+
+describe "Normalize: case" do
+  it "normalizes select with call" do
+    assert_expand "select; when foo; body; when bar; baz; end", <<-CODE
+      __temp_1, __temp_2 = ::Channel.select({foo_select_action, bar_select_action})
```

Comment:
I asked a bit around whether there are existing terms or concepts to this problem.

One interesting note was that it could be a modled as a priority queue with dynamic weights. The weight would denote the readiness state, the queue would hold the operations and pop once. We would need then need to implement a constrained pop that blocks until an element with a high enough weight is available. But all of this doesn't help much with the name for the containers in the queue.

The other note I received was that OCaml calls this "alternatives", or its formal parent CSP calls it "choice". https://en.wikipedia.org/wiki/Communicating_sequential_processes#Algebraic_operators
I think that could actually work here as a generalized construct, `Choice.select`, `Channel#receive_choice`, `Channel#send_choice`, `Choice#ready?`, `Choice#execute`, `Choice#wait`, `Choice#unwait`.

I'm not sure I follow you on the runtime internals, `Choice` would define the the interface with how the operations have to behave, things like "`wait` should ensure the current fiber is rescheduled when value of `ready?` changes" and "`unwait` should remove any reschedule setup by `wait`". We then just to document `Scheduler.enqueue` and that's it, the rest is up to the implementer.